apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/stu3/providers/PatientResourceProvider.java

```diff
@@ -337,51 +314,65 @@ private CcwCodebookVariable partDCwVariableFor(String system) {
-      query.setMaxResults(pagingArgs.getPageSize());
-    }
+  /**
+   * Fetch beneficiaries for the PartD coverage parameter
+   *
+   * @param coverageId coverage type
+   * @param includedIdentifiers list from the includeIdentifier header
+   * @param paging specified
+   * @return the beneficiaries
+   */
+  private List<Beneficiary> fetchBeneficiaries(
+      TokenParam coverageId, List<String> includedIdentifiers, PatientLinkBuilder paging) {
+    List<SetAttribute<Beneficiary, ?>> withRelations = new LinkedList<>();
+    if (hasHICN(includedIdentifiers)) withRelations.add(Beneficiary_.beneficiaryHistories);
+    if (hasMBI(includedIdentifiers)) withRelations.add(Beneficiary_.medicareBeneficiaryIdHistories);
 
-    return query.getResultList();
-  }
+    String contractMonth =
+        coverageId.getSystem().substring(coverageId.getSystem().lastIndexOf('/') + 1);
+    CcwCodebookVariable partDContractMonth = partDCwVariableFor(contractMonth);
+    SingularAttribute<Beneficiary, String> contractMonthField = partDFieldFor(partDContractMonth);
+    String contractCode = coverageId.getValueNotNull();
 
-  private CriteriaQuery queryBeneficiariesBy(
-      SingularAttribute<Beneficiary, String> field, String value) {
-    List<SetAttribute<Beneficiary, ?>> withRelations =
-        new LinkedList<SetAttribute<Beneficiary, ?>>();
-    return queryBeneficiariesBy(field, value, withRelations);
+    CriteriaQuery<Beneficiary> criteria =
+        queryBeneficiariesBy(contractMonthField, contractCode, paging, withRelations);
+    return entityManager.createQuery(criteria).setMaxResults(paging.getPageSize()).getResultList();
   }
 
-  private CriteriaQuery queryBeneficiariesBy(
+  /**
+   * Build a criteria for a general Beneficiary query
+   *
+   * @param field to match on
+   * @param value to match on
+   * @param paging to use for the result set
+   * @param joins to add for many-to-one relations
+   * @return the criteria
+   */
+  private CriteriaQuery<Beneficiary> queryBeneficiariesBy(
       SingularAttribute<Beneficiary, String> field,
       String value,
-      List<SetAttribute<Beneficiary, ?>> relations) {
-
-    CriteriaBuilder builder = entityManager.getCriteriaBuilder();
-    CriteriaQuery<Beneficiary> beneMatches = builder.createQuery(Beneficiary.class);
-    Root<Beneficiary> beneMatchesRoot = beneMatches.from(Beneficiary.class);
-    relations.stream()
-        .forEach(
-            f -> {
-              beneMatchesRoot.fetch(f, JoinType.LEFT);
-            });
-    beneMatches.select(beneMatchesRoot);
-    beneMatches.where(builder.equal(beneMatchesRoot.get(field), value));
-
-    return beneMatches;
+      PatientLinkBuilder paging,
+      List<SetAttribute<Beneficiary, ?>> joins) {
+
+    CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+    CriteriaQuery<Beneficiary> criteria = cb.createQuery(Beneficiary.class);
+    Root<Beneficiary> b = criteria.from(Beneficiary.class);
```

Comment:
For readability, I'm trying make this code follow SQL somewhat. `b` is used as a table alias would be used in a SQL statement.

Reply:
ty this is much cleaner